Documentation/man/man3/Haar.3
.TH "Haar" 3 "Fri Mar 17 2017" "Version 1" "targeter" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Haar
.SH SYNOPSIS
.br
.PP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBHaarEnergy\fP (cv::Mat data, cv::Mat &energyImage, int width, int height, int iterations, bool laplace=false, bool bSquare=true)"
.br
.ti -1c
.RI "static cv::Mat \fBPyramidLevels\fP (cv::Mat data, int width, int height, int iterations, const int NoFocusImages)"
.br
.ti -1c
.RI "static cv::Mat \fBLaplacianPyramid\fP (cv::Mat im, int levels, bool includeScale=true)"
.br
.ti -1c
.RI "static void \fBHaar2\fP (cv::Mat &data, int iterations)"
.br
.ti -1c
.RI "template<class T > static T * \fBPyramidLevels\fP (T *data, int width, int height, int iterations, int NoFocusImages)"
.br
.ti -1c
.RI "template<class T > static T * \fBPyramidRegionLevels\fP (T *data, int width, int height, int iterations, int NoFocusImages)"
.br
.ti -1c
.RI "template<class T > static void \fBHaar1\fP (T *data, int n)"
.br
.ti -1c
.RI "template<class T > static void \fBHaar2\fP (T *data, int width, int height, int iterations, bool AbsImage=false)"
.br
.ti -1c
.RI "template<class T > static void \fBiHaar2\fP (T *data, int width, int height, int iterations)"
.br
.ti -1c
.RI "template<class T > static void \fBiHaar1\fP (T *data, int n)"
.br
.ti -1c
.RI "template<class T > static void \fBiHaar2\fP (cv::Mat &data, int iterations)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 18 of file haar\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "template<class T > static void Haar::Haar1 (T * data, int n)\fC [inline]\fP, \fC [static]\fP"
Perform 1D \fBHaar\fP wavelet transform
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::Haar1\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * data 
.br
\fIint\fP n 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 259 of file haar\&.h\&.
.SS "void Haar::Haar2 (cv::Mat & data, int iterations)\fC [static]\fP"
Perform 2D \fBHaar\fP wavelet pyramid transform with OpenCV image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::Haar2\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP & data 
.br
\fIint\fP iterations (no of levels) 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 214 of file haar\&.cpp\&.
.SS "template<class T > static void Haar::Haar2 (T * data, int width, int height, int iterations, bool AbsImage = \fCfalse\fP)\fC [inline]\fP, \fC [static]\fP"
Perform 1D \fBHaar\fP wavelet transform on 1D image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::Haar2\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * data 
.br
\fIint\fP width 
.br
\fIint\fP height 
.br
\fIint\fP iterations 
.br
\fIbool\fP AbsImage 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 297 of file haar\&.h\&.
.SS "void Haar::HaarEnergy (cv::Mat data, cv::Mat & energyImage, int width, int height, int iterations, bool laplace = \fCfalse\fP, bool bSquare = \fCtrue\fP)\fC [static]\fP"
Get energy image (summed coefficient magnitude) from haar wavelet pyramid
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::HaarEnergy\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP data 
.br
\fIcv::Mat\fP & energyImage 
.br
\fIint\fP width 
.br
\fIint\fP height 
.br
\fIint\fP iterations 
.br
\fIbool\fP laplace 
.br
\fIbool\fP bSquare 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 29 of file haar\&.cpp\&.
.SS "template<class T > static void Haar::iHaar1 (T * data, int n)\fC [inline]\fP, \fC [static]\fP"
Perform inverse \fBHaar\fP wavelet transform on 1D image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::iHaar1\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * data 
.br
\fIint\fP n 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 422 of file haar\&.h\&.
.SS "template<class T > static void Haar::iHaar2 (T * data, int width, int height, int iterations)\fC [inline]\fP, \fC [static]\fP"
Perform inverse \fBHaar\fP wavelet transform on 2D image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::iHaar2\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * data 
.br
\fIint\fP width 
.br
\fIint\fP height 
.br
\fIint\fP iterations 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 363 of file haar\&.h\&.
.SS "template<class T > static void Haar::iHaar2 (cv::Mat & data, int iterations)\fC [inline]\fP, \fC [static]\fP"
Perform inverse \fBHaar\fP wavelet transform on 2D OpenCV image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::iHaar2\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP & data 
.br
\fIint\fP iterations 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access private 
.RE
.PP

.PP
Definition at line 455 of file haar\&.h\&.
.SS "cv::Mat Haar::LaplacianPyramid (cv::Mat im, int levels, bool includeScale = \fCtrue\fP)\fC [static]\fP"
perform Laplacian pyramid
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::LaplacianPyramid\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP im 
.br
\fIint\fP levels 
.br
\fIbool\fP includeScale 
.RE
.PP
\fBReturns:\fP
.RS 4
cv::Mat Access private 
.RE
.PP

.PP
Definition at line 81 of file haar\&.cpp\&.
.SS "cv::Mat Haar::PyramidLevels (cv::Mat data, int width, int height, int iterations, const int NoFocusImages)\fC [static]\fP"
Get OpenCV image of best (modal) values from Pyramid composition
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::PyramidLevels\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP data 
.br
\fIint\fP width 
.br
\fIint\fP height 
.br
\fIint\fP iterations 
.br
\fIconst\fP int NoFocusImages 
.RE
.PP
\fBReturns:\fP
.RS 4
cv::Mat Access private 
.RE
.PP

.PP
Definition at line 139 of file haar\&.cpp\&.
.SS "template<class T > static T* Haar::PyramidLevels (T * data, int width, int height, int iterations, int NoFocusImages)\fC [inline]\fP, \fC [static]\fP"
Get image of best (modal) values from Pyramid composition
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::PyramidLevels\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * data 
.br
\fIint\fP width 
.br
\fIint\fP height 
.br
\fIint\fP iterations 
.br
\fIint\fP NoFocusImages 
.RE
.PP
\fBReturns:\fP
.RS 4
T* Access private 
.RE
.PP

.PP
Definition at line 44 of file haar\&.h\&.
.SS "template<class T > static T* Haar::PyramidRegionLevels (T * data, int width, int height, int iterations, int NoFocusImages)\fC [inline]\fP, \fC [static]\fP"
Get image of best (modal) values from local 8 connected neighbours of Pyramid composition
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName \fBHaar::PyramidRegionLevels\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * data 
.br
\fIint\fP width 
.br
\fIint\fP height 
.br
\fIint\fP iterations 
.br
\fIint\fP NoFocusImages 
.RE
.PP
\fBReturns:\fP
.RS 4
T* Access private 
.RE
.PP

.PP
Definition at line 122 of file haar\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for targeter from the source code\&.


haar.cpp
#include "FocusStack.h"
#include <cmath>
#include <QObject>

#include "opencv2/opencv.hpp"
#include "opencv/highgui.h"
#include "haar.h"


/**
*
* Get energy image (summed coefficient magnitude) from haar wavelet pyramid
*
* @author    David Watts
* @since     2017/03/08
*
* FullName   Haar::HaarEnergy
* Qualifier
* @param     cv::Mat data
* @param     cv::Mat & energyImage
* @param     int width
* @param     int height
* @param     int iterations
* @param     bool laplace
* @param     bool bSquare
* @return    void
* Access     private
*/
void Haar::HaarEnergy(cv::Mat data, cv::Mat &energyImage, int width, int height, int iterations, bool laplace, bool bSquare)
{
	int ii, ij, oi, oj;

	for (int i = 0; i < width; i++)
	{
		for (int j = 0; j < height; j++)
		{
			for (int k = 0; k < iterations; k++)
			{
				ii = i >> (k + 1);
				ij = j >> (k + 1);

				oi = width >> (k + 1);
				oj = height >> (k + 1);

				if (bSquare)
					energyImage.ptr<double>(j)[i] += ((double)data.ptr<double>(oj + ij)[oi + ii] * data.ptr<double>(oj + ij)[oi + ii]) / (double)(1 << k); // DXDY
				else
					energyImage.ptr<double>(j)[i] += fabs((double)data.ptr<double>(oj + ij)[oi + ii] / (double)(1 << k)); // DXDY

				if (!laplace)
				{
					if (bSquare)
						energyImage.ptr<double>(j)[i] += ((double)data.ptr<double>(oj + ij)[ii] * data.ptr<double>(oj + ij)[ii]) / (double)(1 << k);    // SXDY
					else
						energyImage.ptr<double>(j)[i] += fabs((double)data.ptr<double>(oj + ij)[ii] / (double)(1 << k));    // SXDY
					if (bSquare)
						energyImage.ptr<double>(j)[i] += ((double)data.ptr<double>(ij)[oi + ii] * data.ptr<double>(ij)[oi + ii]) / (double)(1 << k);    // SYDX
					else
						energyImage.ptr<double>(j)[i] += fabs((double)data.ptr<double>(ij)[oi + ii] / (double)(1 << k));    // SYDX
				}
			}
		}
	}
}

/**
*
* perform Laplacian pyramid
*
* @author    David Watts
* @since     2017/03/08
*
* FullName   Haar::LaplacianPyramid
* Qualifier
* @param     cv::Mat im
* @param     int levels
* @param     bool includeScale
* @return    cv::Mat
* Access     private
*/
cv::Mat Haar::LaplacianPyramid(cv::Mat im, int levels, bool includeScale)
{
	int l;
	cv::Mat imOut = cv::Mat(im.rows * 2, im.cols * 2, im.type(), cv::Scalar(0));

	cv::Mat currentImg = im;

	for (l = 0; l<levels; l++) {
		cv::Mat down, up;

		cv::pyrDown(currentImg, down);

		cv::pyrUp(down, up, currentImg.size());

		cv::Mat lap = currentImg - up;

		int y = imOut.rows >> (l + 1);
		int x = imOut.cols >> (l + 1);

		cv::Rect r = cv::Rect(x, y, lap.cols, lap.rows);

		lap.copyTo(imOut(r));

		currentImg = down;
	}

	if (includeScale)
	{
		int y = imOut.rows >> (levels + 1);
		int x = imOut.cols >> (levels + 1);

		cv::Rect r = cv::Rect(x, y, currentImg.cols, currentImg.rows);

		currentImg.copyTo(imOut(r));
	}
	return imOut;
}


	

/**
*
* Get OpenCV image of best (modal) values from Pyramid composition
*
* @author    David Watts
* @since     2017/03/08
*
* FullName   Haar::PyramidLevels
* Qualifier
* @param     cv::Mat data
* @param     int width
* @param     int height
* @param     int iterations
* @param     const int NoFocusImages
* @return    cv::Mat
* Access     private
*/
cv::Mat Haar::PyramidLevels(cv::Mat data, int width, int height, int iterations, const int NoFocusImages)
{
	
	int ii, ij, oi, oj, c, x, y, lev;

	int* valueArray = new int[NoFocusImages];

	cv::Mat bestlevels = cv::Mat(height, width, CV_8SC1, cv::Scalar(0));

	for (int i = 0; i < width; i++)
	{
		for (int j = 0; j < height; j++)
		{
			memset(valueArray, 0, sizeof(int)*NoFocusImages);

			for (int k = 0; k < iterations; k++)   // index into wavelet image
			{
				ii = i >> (k + 1);
				ij = j >> (k + 1);

				oi = width >> (k + 1);
				oj = height >> (k + 1);

				if (data.ptr<char>(oj + ij)[oi + ii] >= 0 && data.ptr<char>(oj + ij)[oi + ii] < NoFocusImages)
					valueArray[data.ptr<char>(oj + ij)[oi + ii]]++; // DXDY

				if (data.ptr<char>(oj + ij)[ii] >= 0 && data.ptr<char>(oj + ij)[ii] < NoFocusImages)
					valueArray[data.ptr<char>(oj + ij)[ii]]++;    // SXDY

				if (data.ptr<char>(ij)[oi + ii] >= 0 && data.ptr<char>(ij)[oi + ii] < NoFocusImages)
					valueArray[data.ptr<char>(ij)[oi + ii]]++;    // SYDX

			}

			int bestLevel = 0;
			// get maximum value in value histogram (more frequent level value)
			for (int k = 0; k < NoFocusImages; k++)
			{
				c = valueArray[k];

				if (c > bestLevel)
				{
					bestLevel = c;
					lev = k;
				}
			}
			// should at least count one observation
			if (bestLevel>0)
				bestlevels.ptr<char>(j)[i] = lev;
			else
				bestlevels.ptr<char>(j)[i] = -1;
		}
	}

	delete[] valueArray;

	return bestlevels;
}

	

/**
*
*  Perform 2D Haar wavelet pyramid transform with OpenCV image
*
* @author    David Watts
* @since     2017/03/08
*
* FullName   Haar::Haar2
* Qualifier
* @param     cv::Mat & data
* @param     int iterations (no of levels)
* @return    void
* Access     private
*/
void Haar::Haar2(cv::Mat& data, int iterations)
{
	bool zeroSmooth = false;

	if (data.channels() == 1)
	{
		int width = data.cols;
		int height = data.rows;

		double* row;
		double* col;

		for (int k = 0; k < iterations; k++)
		{
			int levHeight = height >> k;
			int levWidth = width >> k;

			row = new double[levWidth];

			for (int j = 0; j < levHeight; j++)
			{
				for (int i = 0; i < levWidth; i++)
					row[i] = (double)data.ptr<double>(j)[i];

				Haar::Haar1(row, levWidth);

				for (int i = 0; i < levWidth; i++)
					data.ptr<double>(j)[i] = row[i];
			}

			col = new double[levHeight];

			for (int i = 0; i < levWidth; i++)
			{
				for (int j = 0; j < levHeight; j++)
					col[j] = (double)data.ptr<double>(j)[i];

				Haar::Haar1(col, levHeight);

				for (int j = 0; j < levHeight; j++)
					data.ptr<double>(j)[i] = col[j];
			}

			delete[] row;
			delete[] col;
		}

		if (zeroSmooth)
		{
			int levHeight = height >> (iterations);
			int levWidth = width >> (iterations);

			for (int j = 0; j < levHeight + 3; j++)
				for (int i = 0; i < levWidth + 3; i++)
					data.ptr<double>(j)[i] = 0;

		}
	}
	else
	{
		std::cout << " " << __FUNCTION__ << ": error processing image with more than one channel" << std::endl;
	}
}



/*
void LaplacianEnergy(QList<Mat>& arrIM, Mat& smallestLevel, int levels)
{
foreach(Mat im , arrIM) {
int valueArray[NoFocusImages];

Mat bestlevels = cv::Mat(height, width, CV_8SC1, cv::Scalar(0));

for (int i = 0; i < width; i++)
{
for (int j = 0; j < height; j++)
{
memset(valueArray, 0, sizeof(int)*NoFocusImages);

// bottom left SYDX
indi   = i;           indj   = j;
startx = indi - fs1;   starty = indj - fs1;
endx   = indi + fs1;   endy   = indj + fs1;

// center
if(data[indi+indj*width]>=0)
valueArray[data[indi+indj*width]]+=15; // center pixel add equal to surround

// fs1 area
for(x=startx; x<endx; x++)
for(y=starty; y<endy; y++)
{
if(x>=0 && y>=oj && x<oi && y<oj+oj)
{
// surround
if(im.ptr(y)[x]>=0)
valueArray[im.ptr(y)[x]]++;
}
}

int bestLevel = 0;
// get maximum value in value histogram (more frequent level value)
for (int k = 0; k < NoFocusImages; k++)
{
c = valueArray[k];

if(c > bestLevel)
{
bestLevel = c;
lev = k;
}
}
// should at least count one observation
if(bestLevel>0)
bestlevels.ptr<char>(j)[i] = lev;
else
bestlevels.ptr<char>(j)[i] = -1;
}
}

return bestlevels;

}
*/